File: Java/DynamicProgramming/DynamicProgrammingOnSubsequences/416-partition-equal-subset-sum/partition-equal-subset-sum.java
class Solution {
    public boolean canPartition(int[] nums) {
        int totalSum = 0;
        int n = nums.length;
        for(int i=0;i<n;i++){
            totalSum += nums[i];
        }
        if(totalSum %2 != 0)return false;

        int k = totalSum/2;
        
        boolean prev[] = new boolean[k+1];        
        prev[0] = true;
        if(nums[0] <= k){
            prev[nums[0]] = true;
        }

        for(int index=1;index<n;index++){
            boolean curr[] = new boolean[k+1];
            for(int target= 1; target<=k ;target++){
                boolean nottake = prev[target];
                boolean take = false;
                if(target >= nums[index]){
                    take = prev[target - nums[index]];
                }

                curr[target] = (take || nottake);
            }
            prev = curr;
        }
        return prev[k];
    }

}

File: Python/DynamicProgramming/DynamicProgrammingOnSubsequences/416-partition-equal-subset-sum/partition-equal-subset-sum.py
class Solution:
    def f_space_optimisation(self, arr,k):
        n = len(arr)
        prev = [False]*(k+1)
        curr = [False]*(k+1)

        prev[0] = True
        curr[0] = True

        if arr[0] <=k:
            prev[arr[0]] = True
        
        for ind in range(1,n):
            curr = [False]*(k+1)
            curr[0] =True
            for target in range(1,k+1):
                not_take = prev[target]
                take = False

                if target >= arr[ind]:
                    take = prev[target -arr[ind]]

                curr[target] = take or not_take

            prev = curr

        return prev[k]
    def canPartition(self, nums: List[int]) -> bool:
        n = len(nums)
        sum =0
        for i in range(n):
            sum += nums[i]

        if sum%2== 0:
            return self.f_space_optimisation(nums, sum//2)
        else:
            return False
        